File: 523.py
# Continuous subarray sum
# MEDIUM

from typing import List

class Solution:
  def checkSubarraySum(self, nums: List[int], k: int) -> bool:
    prefixSum = 0
    prefixToIndex = {0: -1}

    for i, num in enumerate(nums):
      prefixSum += num
      prefixSum %= k
      if prefixSum in prefixToIndex:
        if i - prefixToIndex[prefixSum] > 1:
            return True
      else:
        prefixToIndex[prefixSum] = i

    return False

File: 523.cpp
// Contiguous Subarray Sum
// MEDIUM

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        int prefix = 0;
        for (int i = 0; i < nums.size(); i ++) {
            prefix += nums[i];
            if (prefix % k == 0) {
                if (i > 0) return true;
            }

            if (mp.find(prefix % k) != mp.end()) {
                if ((i - mp[prefix % k]) >= 2) {
                    return true;
                }
            } else {
                mp[prefix % k] = i;
            }
        }

        return false;
    }
};
